compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -113,7 +121,218 @@ static void specializeGenericTransposeOp(RewriterBase &rewriter,
+//   indexing_maps = [affine_map<(d0, d1, d2) -> (d2, d0, d1)>,
+//                    affine_map<(d0, d1, d2) -> (d0, d1, d2)>]
+//   ins(%0 : tensor<2x7x5>) outs(%3 : tensor<7x5x2>)
+class BubbleTransposeThroughDpsInit
+    : public OpRewritePattern<linalg::TransposeOp> {
+public:
+  using OpRewritePattern<linalg::TransposeOp>::OpRewritePattern;
+  BubbleTransposeThroughDpsInit(MLIRContext *ctx, bool aggressiveProp,
+                                PatternBenefit b = 1)
+      : OpRewritePattern<linalg::TransposeOp>(ctx, b),
+        allowGeneralizing(aggressiveProp) {}
+
+  LogicalResult matchAndRewrite(linalg::TransposeOp transposeOp,
+                                PatternRewriter &rewriter) const override {
+    if (!IREE::Flow::isNonNullAndOutsideDispatch(transposeOp)) {
+      return failure();
+    }
+    OpResult result = dyn_cast<OpResult>(transposeOp.getInput());
+    if (!result) {
+      return failure();
+    }
+    if (!result.hasOneUse()) {
+      return failure();
+    }
+    auto linalgOp = dyn_cast<linalg::LinalgOp>(result.getOwner());
+    if (!linalgOp) {
+      return failure();
+    }
+
+    int64_t resultIndex = result.getResultNumber();
+    auto maybeGenericOp = getGenericOpOrGeneralizeContraction(
+        rewriter, result.getOwner(), allowGeneralizing);
+    if (failed(maybeGenericOp)) {
+      return failure();
+    }
+    auto genericOp = maybeGenericOp.value();
+    result = genericOp->getOpResult(resultIndex);
+
+    ArrayRef<int64_t> perm = transposeOp.getPermutation();
+    auto invPerm = invertPermutationVector(perm);
+
+    OpOperand *init = genericOp.getDpsInitOperand(resultIndex);
+    SmallVector<Value> inits = genericOp.getDpsInits();
+    Value newInit = createTranspose(rewriter, init->get(), perm);
+    inits[resultIndex] = newInit;
+
+    SmallVector<Type> resultTypes(genericOp->getResultTypes());
+    resultTypes[resultIndex] = newInit.getType();
+
+    SmallVector<AffineMap> newIndexingMaps = genericOp.getIndexingMapsArray();
+    AffineMap resultMap =
+        newIndexingMaps[genericOp.getNumDpsInputs() + resultIndex];
+    SmallVector<AffineExpr> newExprs =
+        applyPermutation(resultMap.getResults(), perm);
+    AffineMap transposedMap =
+        AffineMap::get(resultMap.getNumDims(), resultMap.getNumSymbols(),
+                       newExprs, rewriter.getContext());
+    newIndexingMaps[genericOp.getNumDpsInputs() + resultIndex] = transposedMap;
+
+    auto newGenericOp = rewriter.create<linalg::GenericOp>(
+        genericOp.getLoc(), resultTypes, genericOp.getDpsInputs(), newInit,
+        newIndexingMaps, genericOp.getIteratorTypesArray(),
+        /*bodyBuild=*/nullptr, linalg::getPrunedAttributeList(genericOp));
+    rewriter.cloneRegionBefore(genericOp.getRegion(), newGenericOp.getRegion(),
+                               newGenericOp.getRegion().begin());
+
+    // Remap iteration space of the generic to be in order of the output.
+    SmallVector<unsigned int> interchange;
```

Comment:
Can you do this before you create the op instead of after?

Reply:
Using the interchange helper also lets it update `linalg.index` ops too, and it updates the generic in-place, so I figured it would be less code duplication to do it like this. Because of the way the interchange is constructed, it shouldn't fail so I changed it to an assert and added some comments. If you think it's better to do it before I can still make this change though.